controllers/hyperparametertuningjob/hyperparametertuningjob_controller.go

```diff
@@ -19,437 +19,292 @@ package hyperparametertuningjob
-		return RequeueAfterInterval(r.PollInterval, nil)
+	if _, err := ctx.SageMakerClient.CreateHyperParameterTuningJob(ctx, &createTuningJobInput); err != nil {
+		return errors.Wrap(err, "Unable to create HyperParameter Tuning Job")
 	}
 
-	if observedStatus == completed || observedStatus == stopped || observedStatus == failed {
-		return NoRequeue()
-	}
-	// norequeue for unknown status
-	ctx.Log.Info("Job is in unknown status, no requeue")
-	return NoRequeue()
+	return nil
 }
 
-func (r *HyperparameterTuningJobReconciler) updateJobStatus(ctx reconcileRequestContext, desiredStatus hpojobv1.HyperparameterTuningJobStatus) error {
-	ctx.Log.Info("Updating job status", "new-status", desiredStatus)
+// Clean up all models and endpoints, then remove the HostingDeployment finalizer.
+func (r *Reconciler) cleanupAndRemoveFinalizer(ctx reconcileRequestContext) error {
+	var err error
 
-	// Perform deep copy so as to avoid side-effect in ctx.
-	root := ctx.Job.DeepCopy()
-	root.Status = desiredStatus
+	if err = ctx.HPOTrainingJobSpawner.DeleteSpawnedTrainingJobs(ctx, *ctx.TuningJob); err != nil {
+		return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Not all associated TrainingJobs jobs were deleted"))
+	}
 
-	if err := r.Status().Update(ctx, root); err != nil {
-		ctx.Log.Error(err, "Error updating job status", "job", root)
-		return err
+	ctx.TuningJob.ObjectMeta.Finalizers = controllers.RemoveString(ctx.TuningJob.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
+	if err = r.Update(ctx, ctx.TuningJob); err != nil {
+		return errors.Wrap(err, "Failed to remove finalizer")
 	}
+	ctx.Log.Info("Finalizer has been removed")
 
 	return nil
 }
 
-func (r *HyperparameterTuningJobReconciler) statusMatchesDescription(ctx reconcileRequestContext) bool {
-	// TODO implement
-	return false
+// If job is running, stop it and requeue
+// If job is stopping, update status and requeue
+// If job has finished, failed or stopped remove finalizer and don't requeue
+func (r *Reconciler) deleteHyperparameterTuningJob(ctx reconcileRequestContext) error {
+	if _, err := ctx.SageMakerClient.StopHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil {
+		return err
+	}
+	return nil
 }
 
-// Extract the BestTrainingJob from the SageMaker description if it exists and convert it to our Kubernetes type.
-func (r *HyperparameterTuningJobReconciler) createBestTrainingJob(ctx reconcileRequestContext) *commonv1.HyperParameterTrainingJobSummary {
-
-	if ctx.SageMakerDescription.BestTrainingJob == nil {
-		ctx.Log.Info("Cannot get BestTrainingJob: No BestTrainingJob in HPO description")
-		return nil
+// Clean up all models and endpoints, then remove the HostingDeployment finalizer.
+func (r *Reconciler) addBestTrainingJobToStatus(ctx reconcileRequestContext) error {
+	if ctx.TuningJobDescription.BestTrainingJob == nil {
+		return errors.New("No BestTrainingJob in HPO description")
```

Comment:
It seems like a bad idea to fail hard here. The BestTrainingJob will not always be present.